File: webapp/api/KeyInfoService.ts
import * as When from 'when';
import * as $ from 'jquery';
import {IApiCaller} from './IApiCaller';


export interface IKeyInfoService {
    AddKey(keyid: number, vcode: string): When.Promise<number>;
    GetAll(): When.Promise<ts.dto.KeyInfoDto[]>;
    Delete(keyinfoid: number): When.Promise<void>;
}

export interface IKeyInfoServiceContext {
    keyInfoService: IKeyInfoService;
}

export class KeyInfoService implements IKeyInfoService {
    _api: IApiCaller;

    constructor(api: IApiCaller) {
        this._api = api;
    }

    public AddKey(keyid: number, vcode: string): When.Promise<number> {
        var r = this._api.post<ts.dto.SingleLongDto>("/api/keyinfo/add", { KeyId: keyid, VCode: vcode });
        return r.then((data: ts.dto.SingleLongDto) => data.value);
    }

    public GetAll(): When.Promise<ts.dto.KeyInfoDto[]> {
        return this._api.get<ts.dto.KeyInfoDto[]>("/api/keyinfo");
    }

    public Delete(keyinfoid: number): When.Promise<void> {
        var r = this._api.post<void>("/api/keyinfo/delete", { KeyInfoId: keyinfoid});
        return r;
    }
}

File: api/KeyInfoService.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Runtime.Remoting.Messaging;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Nancy;
using Nancy.ModelBinding;
using Nancy.Routing;
using Serilog.Enrichers;
using ts.data;
using ts.dto;
using ts.services;

namespace ts.api
{
    public class KeyInfoService: NancyModule
    {
        private readonly IKeyInfoService _keyInfoService;

        public KeyInfoService(IKeyInfoService keyInfoService)
        {
            _keyInfoService = keyInfoService;
            Post["/keyinfo/add", runAsync:true] = async (_, ct) => await Add();
            Post["/keyinfo/delete", runAsync: true] = async (_, ct) => await DeleteKey();
            Get["/keyinfo", runAsync: true] = async (_, ct) => await GetAll();
        }

        struct AddModel
        {
            public long KeyId { get; set; }
            public string VCode { get; set; }
        }

        private async Task<SingleLongDto> Add()
        {
            var m = this.Bind<AddModel>();

            var val = await _keyInfoService.Add((long) this.Context.Request.Session["UserId"], m.KeyId, m.VCode);
            return new SingleLongDto() {Value = val};
        }

        struct DeleteModel
        {
            public long KeyInfoId { get; set; }
        }

        private async Task<Nancy.HttpStatusCode> DeleteKey()
        {
            var m = this.Bind<DeleteModel>();
            await _keyInfoService.Delete(m.KeyInfoId);

            return Nancy.HttpStatusCode.OK;
        }

        private async Task<List<KeyInfoDto>> GetAll()
        {
            var userid = (long) this.Context.Request.Session["UserId"];
            return await _keyInfoService.Get(userid);
        }
    }

}
